### PR TITLE
UE

### PR DESCRIPTION
**Pull request by una-dinosauria**

### CODE REVIEW

File: data/headset_dataset.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+            SingleCaptureDataset(cap, cap_dir, code_dir, downsample, time_downsample, cameras_specified)
+            for cap, cap_dir, code_dir in zip(captures, directories, latent_code_directories)
+        ]
+
+    def __iter__(self):
+        while True:
+            dataset_idx = np.random.randint(0, len(self.datasets))
+            # dataset_idx = torch.randint(0, len(self.datasets), (1,)).item()
+            dataset = self.datasets[dataset_idx]
+            yield dataset.get_random_frame()
```
Comment (una-dinosauria) [inline data/headset_dataset.py:220]: This is doing stratified sampling, which ideally should be done by a `DatasetSampler` on top of a dataset with random access. Could we at least leave a note with a TODO to improve this at a later stage?

File: data/headset_dataset.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+
+if __name__ == "__main__":
+    import sys
```
Comment (una-dinosauria) [inline data/headset_dataset.py:224]: Do we need this main? If not, we should remove it

File: ddp-train-ue.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+handler.setFormatter(formatter)
+root.addHandler(handler)
+
+torch.backends.cudnn.benchmark = True  # gotta go fast!
+
+
+def gen_optimizer(
+    net: torch.nn.Module,
+    optim_type: str,
+    batchsize: int,
```
Comment (una-dinosauria) [inline ddp-train-ue.py:67]: Can we refactor this with ddp-train.py?

File: ddp-train-ue.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+            "optimizer": optim_type,
+        }
+
+        tensorboard_logger.add_hparams(tb_hyperparams, {"hp_metric": 1.0}, run_name=".")
+    return params, opt
+
+
+def setup(local_rank, world_rank, world_size, master_address: str, master_port: str):
+    logging.info(f"{local_rank=} {world_rank=} {world_size=} {master_address=}, {master_port=}")
+
```
Comment (una-dinosauria) [inline ddp-train-ue.py:97]: ibid

File: ddp-train-ue.py [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+                    tensorboard_logger.add_scalar(k, float(v.mean().item()), iternum)
+
+        del cudadata
+    cleanup()
+
+
+if __name__ == "__main__":
+    __spec__ = None  # to use ipdb
+
+    # parse arguments
```
Comment (una-dinosauria) [inline ddp-train-ue.py:334]: remove main if not needed

File: models/headset_encoders/loss.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+        """
+        A wrapper for running, decoding the evaluating universal headset facial encoders
+
+        Args:
+            encoder: The universal decoder to use.
+            identities: The list of decoder identity names to use.
+        """
+        super().__init__()
+        self._encoder = encoder
+        self._decoder = UDWrapper(ud_exp_name="/uca/leochli/oss/ava256_universal_decoder", identities=identities)
```
Comment (una-dinosauria) [inline models/headset_encoders/loss.py:33]: Should the decoder be passed as an argument?

File: models/headset_encoders/universal.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+        return {
+            "expression": expr_code,
+        }
+
+
+if __name__ == "__main__":
+    num_conds = 1
+    hmc_cond_imgs = torch.rand((2, num_views, num_conds, in_chans, 400, 400), dtype=torch.float32)
+    expression_encoder = UniversalEncoder(
```
Comment (una-dinosauria) [inline models/headset_encoders/universal.py:124]: Can we remove this main?

File: tests/test_headset_encoder.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+import torch
+
+from models.headset_encoders.universal import UniversalEncoder
+
+
+def test_encoder_multi_cond():
+    """Tests for the universal headset encoder with different number of conditionings."""
+    in_chans = 1
+    out_chans = 256
```
Comment (una-dinosauria) [inline tests/test_headset_encoder.py:15]: Awesome! Thanks for adding tests! :)

**Review by una-dinosauria — COMMENTED** [review]
Could you add an overall overview of how this code is used?

Ie, what are the inputs, outputs, what should run first, etc?

File: data/headset_dataset.py (hunk 1 above)
Comment (jerrybai1995) [reply data/headset_dataset.py:220]: Currently the dataset is a subclass of the IterableDataset, which I believe doesn't support using samplers. Specifically, we want an infinite looping over the data, rather than epoch-based training.

For that reason, an alternative may be to re-write the dataset as a map-style dataset and then add a sampler with random access and replacement when we define dataloaders. If you think that makes sense, I'll leave a TODO for that path. Please let me know!

File: ddp-train-ue.py (hunk 3 above)
Comment (jerrybai1995) [reply ddp-train-ue.py:67]: I tried but the name "ddp-train" cannot be imported due to the dash "-". Should be possible with importlib, if you think we should use that? (Or we can rename "ddp-train" to "ddp_train"...)

File: data/headset_dataset.py (hunk 2 above)
Comment (jerrybai1995) [reply data/headset_dataset.py:224]: Yeah will do. I'll remove all name=main stuff.

File: ddp-train-ue.py (hunk 4 above)
Comment (jerrybai1995) [reply ddp-train-ue.py:97]: Same as the ddp-train issue above.

File: models/headset_encoders/loss.py (hunk 6 above)
Comment (jerrybai1995) [reply models/headset_encoders/loss.py:33]: Will fix.

File: tests/test_headset_encoder.py (hunk 8 above)
Comment (leochli) [reply tests/test_headset_encoder.py:15]: :)

File: data/headset_dataset.py (hunk 1 above)
Comment (una-dinosauria) [reply data/headset_dataset.py:220]: It makes sense that iterables don't support samplers -- you'd need a random access dataset for that. We can also make an infinite sampler for the non-stop training over a finite number of iterations. A TODO is probably fine for now.

File: ddp-train-ue.py (hunk 3 above)
Comment (una-dinosauria) [reply ddp-train-ue.py:67]: Oh, I was thinking of extracting it to a common 3rd file

File: ddp-train-ue.py (hunk 5 above)
Comment (jerrybai1995) [reply ddp-train-ue.py:334]: This is needed to start the training script on command line.
